java/src/main/java/com/xebia/functional/xef/java/auto/AIScope.java

```diff
@@ -0,0 +1,147 @@
+package com.xebia.functional.xef.java.auto;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.module.jsonSchema.JsonSchema;
+import com.fasterxml.jackson.module.jsonSchema.JsonSchemaGenerator;
+import com.xebia.functional.loom.LoomAdapter;
+import com.xebia.functional.xef.AIError;
+import com.xebia.functional.xef.auto.CoreAIScope;
+import com.xebia.functional.xef.embeddings.Embeddings;
+import com.xebia.functional.xef.embeddings.OpenAIEmbeddings;
+import com.xebia.functional.xef.env.OpenAIConfig;
+import com.xebia.functional.xef.llm.openai.KtorOpenAIClient;
+import com.xebia.functional.xef.llm.openai.LLMModel;
+import com.xebia.functional.xef.llm.openai.functions.CFunction;
+import com.xebia.functional.xef.llm.openai.images.ImageGenerationUrl;
+import com.xebia.functional.xef.llm.openai.images.ImagesGenerationResponse;
+import com.xebia.functional.xef.textsplitters.TextSplitter;
+import com.xebia.functional.xef.vectorstores.LocalVectorStore;
+import com.xebia.functional.xef.vectorstores.VectorStore;
+import kotlin.collections.CollectionsKt;
+import kotlin.jvm.functions.Function1;
+import com.xebia.functional.xef.pdf.PDFLoaderKt;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.List;
+
+public class AIScope implements AutoCloseable {
+    private CoreAIScope coreAIScope;
+    private ObjectMapper om;
+    private JsonSchemaGenerator schemaGen;
+
+    public AIScope(CoreAIScope coreAIScope) {
+        this.coreAIScope = coreAIScope;
+        this.om = new ObjectMapper();
+        this.schemaGen = new JsonSchemaGenerator(om);
+    }
+
+    public AIScope(CoreAIScope coreAIScope, ObjectMapper om) {
+        this.coreAIScope = coreAIScope;
+        this.om = om;
+        this.schemaGen = new JsonSchemaGenerator(om);
+    }
+
+    public static <T> T run(Function1<AIScope, T> block) {
+        OpenAIConfig config = new OpenAIConfig();
+        KtorOpenAIClient client = new KtorOpenAIClient(config);
+        try {
+            Embeddings embeddings = new OpenAIEmbeddings(config, client);
+            VectorStore vectorStore = new LocalVectorStore(embeddings);
+            CoreAIScope scope = new CoreAIScope(LLMModel.getGPT_3_5_TURBO(), LLMModel.getGPT_3_5_TURBO_FUNCTIONS(), client, vectorStore, embeddings, 3, "user", false, 0.4, 1, 20, 500);
+            return block.invoke(new AIScope(scope));
+        } finally {
+            client.close();
+        }
```

Comment:
Maybe we could add here some `try-with-resources` syntax?
```java
try (KtorOpenAIClient client = new KtorOpenAIClient(config)) {
  // the code
}
```

Reply:
Yes, that was my original intention but I need to investigate what is going on under the hood. Java doesn't recognise `KtorOpenAIClient` as implementing Java's `AutoCloseable` but that might've just been IDEA acting up.

Reply:
I've been reading and it seems like, although `AutoCloseable` in Kotlin and Java behave in the same way, they're not the same interface, maybe that's the reason Java didn't recognise `KtorOpenAIClient` as a `AutoCloseable`

Reply:
Aha, it seems you're right but this also seems like a temporary measure 🤔 Normally it's possible for it to be the same type, but it seems it's not the case. I made some updates to the PR that resolves this issue though and you can now use `AIScope` with `try-with-resources` syntax.

https://github.com/JetBrains/kotlin/blob/a9d0ff83cf7dc847213c31b61b7d0c88ab62a95c/libraries/stdlib/jdk7/src/kotlin/AutoCloseableJVM.kt#L20